src/datasets/arrow_dataset.py

```diff
@@ -4666,6 +4666,42 @@ def to_dict(self, batch_size: Optional[int] = None, batched: bool = False) -> Un
                 for offset in range(0, len(self), batch_size)
             )
 
+    def to_list(self, batch_size: Optional[int] = None, batched: bool = False) -> Union[list, Iterator[list]]:
+        """Returns the dataset as a Python list. Can also return a generator for large datasets.
+
+        Args:
+            batched (`bool`):
+                Set to `True` to return a generator that yields the dataset as batches
+                of `batch_size` rows. Defaults to `False` (returns the whole datasets once).
+            batch_size (`int`, *optional*): The size (number of rows) of the batches if `batched` is `True`.
+                Defaults to `datasets.config.DEFAULT_MAX_BATCH_SIZE`.
+
+        Returns:
+            `list` or `Iterator[list]`
+
+        Example:
+
+        ```py
+        >>> ds.to_list()
+        ```
+        """
+        if not batched:
+            return query_table(
+                table=self._data,
+                key=slice(0, len(self)),
+                indices=self._indices if self._indices is not None else None,
+            ).to_pylist()
+        else:
+            batch_size = batch_size if batch_size else config.DEFAULT_MAX_BATCH_SIZE
+            return (
+                query_table(
+                    table=self._data,
+                    key=slice(offset, offset + batch_size),
+                    indices=self._indices if self._indices is not None else None,
+                ).to_pylist()
+                for offset in range(0, len(self), batch_size)
+            )
```

Comment:
You can already use `dataset.iter(batch_size=batch_size)` to get an iterator.

A function named `to_list()` should return a list IMO

Reply:
The same thing here. We could remove these params here and then deprecate them in `to_dict` (to be aligned), but I don't think this is worth the hassle, considering the default behavior is the "expected" one (returning an entire list/dict). Hence, I think we can merge the PR in the current state.

Reply:
Oops, I noticed your 2nd message after replying. I agree this is not the best design (e.g., I wasn't even aware of these parameters before this PR 😄), so let's deprecate these parameters.

@kyoto7250 Let me know if you'd like to address this, or I can do it later this week.

Reply:
@mariosasko 

Of course, I want to try it!

So it means that we should return a batch of 'loaded objects' (not an iterator), correct?
And, should we change both `Dataset.to_dict` and `Dataset.to_list` in this PR?

Reply:
No, it means `to_dict`/`to_list` should always return a dictionary/list, not an iterator. So we need to remove `batch_size` and `batched` in `to_list` and deprecate these parameters in `to_dict` (check https://github.com/huggingface/datasets/pull/5142 to see how we deprecate params in `datasets`).

Reply:
I added two commits.

Could you please tell me in which version this feature will be deprecated and in which version this feature will be removed, for the sake of displaying the message?

Reply:
You can find the versions in my review.